### DISCUSSION TITLE
How to use dnscrypt with tor browser

### DISCUSSION THREAD

**Discussion by Romulo-Romero** [DNSCrypt/dnscrypt-proxy#2636] [Q&A]
Hi everyone,
I have been trying to run to tor browser with dnscrypt, but haven't been able to set it up. The reason I want to use dnscrypt with tor browser is that I've encountered several tor exit nodes utilizing google dns, which feels extremely awkward and not helpful due to censorship which it is involved with google servers.
So the question is: how to setup tor browser to use dnscrypt as its resolve ?
thanks

**Comment by lifenjoiner** [comment]
This is not the right direction. You should find out how the tor thing works first.
1. The onion addresses are not actual DNS names. They are resolved to be 127.x.x.x.
2. The sites must be visited using Tor proxy.

-> https://en.wikipedia.org/wiki/.onion

**Comment by Romulo-Romero** [comment]
@lifenjoiner Hi,
I understand that the 'TLD' onion is not actually resolvable by the central authority, and that the addresses are 'resolved' by the exit nodes, but I just wanted to have the same effect as when I am running the tor expert bundle and have dnscrypt to resolve the addresses except for the onion addresses. 
But thanks for the suggestion and insight, I hope to have a better grasp of how the whole onion infrastructure works.

**Comment by lifenjoiner** [comment]
Seems I didn't get the exact problem, from the email notification with original post but without updates. Explaining the problem more specific is good, especially for the less used case.

For normal websites over onion network, I think it is a proxy first thing. Tor browser adopts socks5. Doing remote DNS resolve is preferred.
The (user) application does a DNS resolve first is called **unsafe** socks. I guess Tor browser has prevented it.

Firefox has the `Proxy DNS when using SOCKS v5`/`network.proxy.socks_remote_dns` option.

**Comment by Romulo-Romero** [comment]
@lifenjoiner hi,
I've tried to set it to false but still I am not able to use dnscrypt. I get the following error : socks connect tcp 127.0.0.1:9050->178.216.201.128:2053: dial tcp 127.0.0.1:9050: connectex: No connection could be made because the target machine actively refused it.
not sure what's going on.
Can you shed some light on this issue ?
